### PR TITLE
Wip cmake

### PR DESCRIPTION
**Pull request by alimaredia**
Lots of features added to CMake, closing the gap between automake!

### CODE REVIEW

**Comment by dx9** [comment]
:+1:  yay!!

**Comment by dx9** [comment]
@alimaredia tried this out today. Noticed it needs rebase because of key-value db source reorg https://github.com/ceph/ceph/commit/e1783d23 from #6312

**Comment by alimaredia** [comment]
@dx9 this pull request should be a quick bandaid: https://github.com/ceph/ceph/pull/6539

I'm working on mimicking the Makefile.am in the kv directory with a CMakeLists.txt for a higher quality fix.

**Comment by dx9** [comment]
@alimaredia cool, thanks man

Also noticed if I disable babeltrace the build fails. Not sure if that is suppose to work or not?

**Comment by alimaredia** [comment]
@dx9 I tried to recreate babeltrace issue, but wasn't able to. What specifically did you run that caused the build to fail?

**Comment by alimaredia** [comment]
@jdurgin is the kinetic-client library (https://github.com/Kinetic/kinetic-cpp-client) going to be a sub-module in the future? 

I cloned it, built it with cmake, and noticed that it doesn't have an install target, so I couldn't integrate it into CMake by running a find_package on it.

I could build it as an external_project, but I wanted your or sages input first.

**Comment by jdurgin** [comment]
@alimaredia don't worry about it, it was just a prototype and wouldn't be the right way to integrate with kinetic at this point. I'd be fine with removing it at this point - it's still in the history if we want to reuse any of it in the future.

**Comment by dx9** [comment]
Hey @alimaredia I think the error should happen on any system missing babeltrace. I'm working on a build for Alpine and it doesn't have babeltrace. The build fails with: 

[ 57%] Building CXX object src/CMakeFiles/librbd_replay_prep.dir/rbd_replay/rbd-replay-prep.cc.o
/build/ceph/src/rbd_replay/rbd-replay-prep.cc:20:35: fatal error: babeltrace/babeltrace.h: No such file or directory
compilation terminated.
src/CMakeFiles/librbd_replay_prep.dir/build.make:62: recipe for target 'src/CMakeFiles/librbd_replay_prep.dir/rbd_replay/rbd-replay-prep.cc.o' failed
make[2]: *** [src/CMakeFiles/librbd_replay_prep.dir/rbd_replay/rbd-replay-prep.cc.o] Error 1
CMakeFiles/Makefile2:2166: recipe for target 'src/CMakeFiles/librbd_replay_prep.dir/all' failed
make[1]: *** [src/CMakeFiles/librbd_replay_prep.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by dx9** [comment]
@alimaredia worked for me. thanks

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -121,13 +121,18 @@ if(HAVE_XIO)
...
 
-if(${WITH_TCMALLOC} AND ${HAVE_GPERFTOOLS})
+# sort out which allocator to use
+if(Tcmalloc_FOUND)
+  set(ALLOC_LIBS tcmalloc)
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free")
-  set(TCMALLOC_LIBS tcmalloc)
   set(TCMALLOC_srcs perfglue/heap_profiler.cc)
-else(${WITH_TCMALLOC} AND ${HAVE_GPERFTOOLS})
+elseif(NOT Tcmalloc_FOUND AND JEMALLOC_FOUND)
```
Comment (cbodley) [inline src/CMakeLists.txt]: this is an elseif, so you shouldn't need another check for Tcmalloc_FOUND, just `elseif(JEMALLOC_FOUND)`

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -566,19 +609,35 @@ set(ceph_mon_srcs
...
+  ${CMAKE_DL_LIBS} ${ALLOC_LIBS})
 install(TARGETS ceph-mon DESTINATION bin)
 
+#external project rocksdb
+if(${WITH_ROCKSDB})
+include(ExternalProject)
+ExternalProject_Add(rocksdb
+  SOURCE_DIR ${CMAKE_SOURCE_DIR}/src/rocksdb
+  CONFIGURE_COMMAND true
+  BINARY_DIR ${CMAKE_SOURCE_DIR}/src/rocksdb
```
Comment (cbodley) [inline src/CMakeLists.txt]: BINARY_DIR should be under CMAKE_BINARY_DIR

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -986,11 +1067,11 @@ if(${WITH_RBD})
...
   set(librbd_replay_srcs
-      rbd_replay/actions.cc
+      rbd_replay/actions.cc 
       rbd_replay/BufferReader.cc
-      rbd_replay/ImageNameMap.cc
-      rbd_replay/PendingIO.cc
-      rbd_replay/rbd_loc.cc
+      rbd_replay/ImageNameMap.cc 
+      rbd_replay/PendingIO.cc 
+      rbd_replay/rbd_loc.cc
```
Comment (cbodley) [inline src/CMakeLists.txt]: unnecessary spaces here

File: src/common/TracepointProvider.h [hunk 4]
```diff
@@ -56,7 +56,7 @@ class TracepointProvider : public md_config_obs_t, boost::noncopyable {
 
   template <const Traits &traits>
   static void initialize(CephContext *cct) {
-#if WITH_LTTNG
+#ifdef WITH_LTTNG
```
Comment (cbodley) [inline src/common/TracepointProvider.h]: shouldn't need to change this header after fixing the `#cmakedefine WITH_LTTNG 1`

File: src/erasure-code/shec/CMakeLists.txt [hunk 5]
```diff
@@ -33,4 +33,72 @@ add_dependencies(ec_shec_generic ${CMAKE_SOURCE_DIR}/src/ceph_ver.h)
...
+  set_target_properties(ec_shec_neon PROPERTIES VERSION 2.0.0 SOVERSION 2
+    COMPILE_FLAGS ${ARM_NEON_FLAGS})
+  install(TARGETS ec_shec_neon DESTINATION lib/erasure-code)
+else(ARM_NEOR OR ARM_NEON2)
+  message(STATUS "Skipping target ec_shec_neon: Architecture not ARM")
+endif(ARM_NEON OR ARM_NEON2)
+
+# ec_shec_sse3
+if(INTEL_SSE)
+  set(SHEC_SSE3_FLAGS "-msse")
```
Comment (cbodley) [inline src/erasure-code/shec/CMakeLists.txt]: all of this logic for SHEC_SSE3_FLAGS is duplicated for JERASURE_SSE3_FLAGS. can we move both of these up to erasure-code/CMakeLists.txt with the rest of the try_compile() stuff? i.e. construct variables for INTEL_SSE3_FLAGS and INTEL_SSE4_FLAGS, and use them for both shec and jerasure

**Pull request closed without merging**